packages/loopring_v3/circuit/Utils/Data.h

```diff
@@ -222,353 +340,371 @@ class Ring
+    ethsnarks::FieldT toAccountID;
+    ethsnarks::FieldT tokenID;
+    ethsnarks::FieldT amount;
+    ethsnarks::FieldT feeTokenID;
+    ethsnarks::FieldT fee;
+    ethsnarks::FieldT validUntil;
+    ethsnarks::FieldT to;
+    ethsnarks::FieldT dualAuthorX;
+    ethsnarks::FieldT dualAuthorY;
+    ethsnarks::FieldT data;
+    ethsnarks::FieldT payerToAccountID;
+    ethsnarks::FieldT payerTo;
+    ethsnarks::FieldT payeeToAccountID;
+    ethsnarks::FieldT type;
 };
 
-static void from_json(const json& j, OnchainWithdrawalBlock& block)
+static void from_json(const json& j, Transfer& transfer)
+{
+    transfer.fromAccountID = ethsnarks::FieldT(j.at("fromAccountID"));
+    transfer.toAccountID = ethsnarks::FieldT(j.at("toAccountID"));
+    transfer.tokenID = ethsnarks::FieldT(j.at("tokenID"));
+    transfer.amount = ethsnarks::FieldT(j["amount"].get<std::string>().c_str());
+    transfer.feeTokenID = ethsnarks::FieldT(j.at("feeTokenID"));
+    transfer.fee = ethsnarks::FieldT(j["fee"].get<std::string>().c_str());
+    transfer.validUntil = ethsnarks::FieldT(j.at("validUntil"));
+    transfer.to = ethsnarks::FieldT(j["to"].get<std::string>().c_str());
+    transfer.dualAuthorX = ethsnarks::FieldT(j["dualAuthorX"].get<std::string>().c_str());
+    transfer.dualAuthorY = ethsnarks::FieldT(j["dualAuthorY"].get<std::string>().c_str());
+    transfer.data = ethsnarks::FieldT(j["data"].get<std::string>().c_str());
+    transfer.payerToAccountID = ethsnarks::FieldT(j.at("payerToAccountID"));
+    transfer.payerTo = ethsnarks::FieldT(j["payerTo"].get<std::string>().c_str());
+    transfer.payeeToAccountID = ethsnarks::FieldT(j.at("payeeToAccountID"));
+    transfer.type = ethsnarks::FieldT(j.at("type"));
+}
+
+class Witness
 {
-    block.exchangeID = ethsnarks::FieldT(j["exchangeID"].get<unsigned int>());
+public:
+    TradeHistoryUpdate tradeHistoryUpdate_A;
+    TradeHistoryUpdate tradeHistoryUpdate_B;
 
-    block.merkleRootBefore = ethsnarks::FieldT(j["merkleRootBefore"].get<std::string>().c_str());
-    block.merkleRootAfter = ethsnarks::FieldT(j["merkleRootAfter"].get<std::string>().c_str());
+    BalanceUpdate balanceUpdateS_A;
+    BalanceUpdate balanceUpdateB_A;
+    AccountUpdate accountUpdate_A;
 
-    block.startHash = ethsnarks::LimbT(j["startHash"].get<std::string>().c_str());
+    BalanceUpdate balanceUpdateS_B;
+    BalanceUpdate balanceUpdateB_B;
+    AccountUpdate accountUpdate_B;
 
-    block.startIndex = ethsnarks::FieldT(j["startIndex"].get<std::string>().c_str());
-    block.count = ethsnarks::FieldT(j["count"].get<std::string>().c_str());
+    BalanceUpdate balanceUpdateA_O;
+    BalanceUpdate balanceUpdateB_O;
+    AccountUpdate accountUpdate_O;
 
-    // Read withdrawals
-    json jWithdrawals = j["withdrawals"];
-    for(unsigned int i = 0; i < jWithdrawals.size(); i++)
-    {
-        block.withdrawals.emplace_back(jWithdrawals[i].get<Loopring::OnchainWithdrawal>());
-    }
-}
+    BalanceUpdate balanceUpdateA_P;
```

Comment:
_P is protocol fee, what is _I for?

Reply:
The account used for the index, already removed on my branch.